criu/mem.c

```diff
@@ -172,17 +175,58 @@ static int generate_iovs(struct pstree_item *item, struct vma_area *vma, struct
 	unsigned long pages[3] = {};
 	int ret = 0;
 
+	static char *zero_page = NULL;
+	static char *remote_page = NULL;
+	int zero = 0;
+	struct iovec  local[2];
+	struct iovec  remote[1];
+	int nread = 0;
+	if (opts.skip_zero_pages && zero_page == NULL) {
+		zero_page = xmalloc(PAGE_SIZE);
+		remote_page = xmalloc(PAGE_SIZE);
+		if (zero_page == NULL || remote_page == NULL) {
+			pr_warn("Can't allocate memory - disabling --skip-zero-pages\n");
+			opts.skip_zero_pages = 0;
+		} else	{
+			memzero(zero_page, PAGE_SIZE);
+			local[0].iov_base = remote_page;
+			local[0].iov_len = PAGE_SIZE;
+			remote[0].iov_base = (void *) 0x0;
+			remote[0].iov_len = PAGE_SIZE;
+		}
+	}
+
 	nr_to_scan = (vma_area_len(vma) - *off) / PAGE_SIZE;
 
 	for (pfn = 0; pfn < nr_to_scan; pfn++) {
 		unsigned long vaddr;
 		unsigned int ppb_flags = 0;
 		int st;
 
-		if (!should_dump_page(vma->e, at[pfn]))
-			continue;
-
 		vaddr = vma->e->start + *off + pfn * PAGE_SIZE;
+		/*
+		 * If should_dump_page() returns true, it means the page is in the dumpees resident memory
+		 * (i.e. bit 63 of the page frame number 'at[pfn]' is set) but it is not the zero-page.
+		 */
+		if (should_dump_page(vma->e, at[pfn])) {
+			if (opts.skip_zero_pages) {
+				remote[0].iov_base = (void*)vaddr;
+				nread = process_vm_readv(item->pid->real, local, 1, remote, 1, 0);
+				if (nread == PAGE_SIZE) {
+					zero = memcmp(zero_page, remote_page, PAGE_SIZE);
+					/*
+					 * If the page contains just zeros we can treat it like the zero page and skip it.
+					 * At restore it will be replaced by a reference to the zero page and COWed if accessed.
+					 */
+					if (zero == 0) {
+						pr_info("Zero page detected at virtual addr = %p\n", (void*)vaddr);
```

Comment:
Using `pr_info()` here is very expensive. Would it be possible to remove it?

Reply:
Sure, I can remove it. I mainly used it during development. Do you think `pr_debug()` would be more appropriate, oder better remove it altogether?

Reply:
Removing it altogether would be better because most CRIU users set verbosity level to debug.